### PR TITLE
Site builder build promise

### PR DESCRIPTION
**Pull request by mbland**
cc: @msecret @rogeruiz

### CODE REVIEW

**Comment by mbland** [comment]
Shoulda mentioned, there's another commit in the pipe to return a `Promise` from `SiteBuilder.build` after this, but #29 and the follow-up commit to #32 for test portability need to go in first.

File: lib/site-builder.js [hunk 1]
```diff
@@ -43,12 +43,12 @@ function SiteBuilder(opts, branch, commandRunner, jekyllHelper,
   }
 }
 
-SiteBuilder.prototype.generateBuildDone = function(done) {
+SiteBuilder.prototype.finishBuild = function() {
   var builder = this,
```
Comment (msecret) [inline lib/site-builder.js:48]: Sorry if I've missed this earlier.

Usual JS convention dictates to use the var name `self` for local scoping `this`. Is that possible here, or would it be too much work?

File: lib/site-builder.js [hunk 2]
```diff
@@ -70,18 +71,31 @@ SiteBuilder.prototype.spawn = function(path, args) {
...
+      finish;
+
+  doBuild = function() {
+    return new Promise(function(resolve) {
+      fs.exists(builder.opts.sitePath, function(exists) {
+        resolve(exists ? builder.syncRepo() : builder.cloneRepo());
+      });
+    })
+    .then(function() { return builder.checkForFile('_config.yml'); })
+    .then(function(useJekyll) {
```
Comment (msecret) [inline lib/site-builder.js:90]: Could this promise chain use a `catch` in case an error occurs?

File: lib/site-builder.js [hunk 3]
```diff
@@ -70,18 +71,31 @@ SiteBuilder.prototype.spawn = function(path, args) {
...
+      return (useJekyll ? builder._buildJekyll() : builder._rsync());
     });
   };
-  this.updateLock.doLockedOperation(doBuild, this.generateBuildDone(done));
+
+  finish = function(err) {
+    return builder.finishBuild()
+      .then(function() {
+        done(err);
```
Comment (msecret) [inline lib/site-builder.js]: Is this `done()` function only called when there's an error? Is there a reason this was kept as a callback rather then being switched to a promise?

Comment (mbland) [inline lib/site-builder.js]: Not exactly, but I've updated both this function and `finishBuild()` in a way that is hopefully more straightforward and readable.

File: lib/site-builder.js (hunk 2 above)
Comment (mbland) [inline lib/site-builder.js:90]: `builder.updateLock.doLockedOperation(doBuild)` will propagate any rejections, and the `catch` block in the return statement at the end of `SiteBuilder.build` will handle the failure case before propagating the rejection further.

File: lib/site-builder.js (hunk 1 above)
Comment (mbland) [inline lib/site-builder.js:48]: It's definitely possible, and certainly not too much work. However, even though this is all server-side code, apparently [`self` has been appropriated for the global context](https://developer.mozilla.org/en-US/docs/Web/API/Window/self), so I'd be hesitant to use it that way here. However, you're the more experienced JS dev here, so if you say I'm being paranoid, I'll happily change it.

I've formerly done `var that = this` (and there are still examples of it in this PR), but I've started trying to use names that are more descriptive of the object, hence the selection of `builder`. However, like I said, if you say I should go with `self` or `that`, I'm happy to do so.

**Pull request merged**